tool/tctl/common/sessions_command.go

```diff
@@ -0,0 +1,126 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/gravitational/kingpin"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/service"
+)
+
+// SessionsCommand implements "tctl sessions" group of commands.
+type SessionsCommand struct {
+	config *service.Config
+
+	// format is the output format (text, json, or yaml)
+	format string
+	// sessionsList implements the "tctl sessions ls" subcommand.
+	sessionsList *kingpin.CmdClause
+	// FromUTC is the start time to use for the range of sessions listed by the recorded session listing command
+	FromUTC string
+
+	// ToUTC is the start time to use for the range of sessions listed by the recorded session listing command
+	ToUTC string
+}
+
+// Initialize allows SessionsCommand to plug itself into the CLI parser
+func (c *SessionsCommand) Initialize(app *kingpin.Application, config *service.Config) {
+	c.config = config
+
+	sessions := app.Command("sessions", "Operate on recorded sessions.")
+	c.sessionsList = sessions.Command("ls", "List recorded sessions.")
+	c.sessionsList.Flag("format", "Output format, 'text', 'json', or 'yaml'").Default(teleport.Text).StringVar(&c.format)
+	c.sessionsList.Flag("from-utc", fmt.Sprintf("Start of time range in which sessions are listed. Format %s", time.RFC3339)).StringVar(&c.FromUTC)
+	c.sessionsList.Flag("to-utc", fmt.Sprintf("End of time range in which sessions are listed. Format %s", time.RFC3339)).StringVar(&c.ToUTC)
+}
+
+// TODO: Move this somewhere more appropriate
+const defaultSearchSessionPageLimit = 50
+
+// TryRun attempts to run subcommands like "sessions ls".
+func (c *SessionsCommand) TryRun(ctx context.Context, cmd string, client auth.ClientI) (match bool, err error) {
+	switch cmd {
+	case c.sessionsList.FullCommand():
+		err = c.ListSessions(client)
+	default:
+		return false, nil
+	}
+	return true, trace.Wrap(err)
+}
+
+//TODO: Deduplicate this logic, same functions defined for tsh
+// ListApps prints the list of recorded sessions.
+func (c *SessionsCommand) ListSessions(tc auth.ClientI) error {
+	fromUTC := time.Unix(0, 0)
+	toUTC := time.Now()
+	var err error
+	if c.FromUTC != "" {
+		fromUTC, err = time.Parse(time.RFC3339, c.FromUTC)
+		if err != nil {
+			return trace.Errorf("parsing session listing start time: %v", err)
```

Comment:
To make the error message more useful:

1. Use `trace.BadParameter` instead of `trace.Errorf`.
2. Include actual value we failed to parse.
3. Say what the expected format is.

In general, we should aim to make our messages look like proper messages, we have a tendency of returning errors that read more like internal debug logs. For example, in this case we should say:

```
Failed to parse session listing start time: expected format XXX, got YYY.
```

Reply:
agreed, except that errors should start with a lower case character and should not end with a period to match the rest of the codebase, when the error is logged the sentence format may not make sense